File: Lab7/app/src/main/java/com/example/lab7/MainActivity.java
package com.example.lab7;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.util.Log;
import android.widget.Button;
import android.widget.SeekBar;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    private int progressRabbit=0;
    private int progressTurtle=0;
    private Button btn_start;
    private SeekBar sb_rabbit, sb_turtle;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        btn_start=findViewById(R.id.btn_start);
        sb_rabbit=findViewById(R.id.sb_rabbit);
        sb_turtle=findViewById(R.id.sb_turtle);

        btn_start.setOnClickListener(view -> {
            btn_start.setEnabled(false);
            progressRabbit=0;
            progressTurtle=0;
            sb_rabbit.setProgress(0);
            sb_turtle.setProgress(0);
            runRabbit();
            runTurtle();
        });
    }
    private final Handler handler=new Handler(Looper.myLooper(), new Handler.Callback() {
        @Override
        public boolean handleMessage(@NonNull Message message) {
            Log.e("Rabbit",String.valueOf(progressRabbit));
            Log.e("Turtle",String.valueOf(progressTurtle));

            if(message.what==1){
                sb_rabbit.setProgress(progressRabbit);
            }else if(message.what==2){
                sb_turtle.setProgress(progressTurtle);
            }

            if(progressRabbit>=100 && progressTurtle<100){
                Toast.makeText(MainActivity.this,"兔子勝",Toast.LENGTH_SHORT).show();
                btn_start.setEnabled(true);
            }else if(progressTurtle>=100 && progressRabbit<100){
                Toast.makeText(MainActivity.this,"烏龜勝",Toast.LENGTH_SHORT).show();
                btn_start.setEnabled(true);
            }
            return false;
        }
    });

    private void runRabbit(){
        new Thread(()->{
           Boolean[] sleepProbability={true,true,false};

           while (progressRabbit<=100 && progressTurtle<100){
               try{
                   Thread.sleep(100);
                   if(sleepProbability[(int)(Math.random()*3)])
                       Thread.sleep(300);

               }catch (InterruptedException e){
                   e.printStackTrace();
               }
               progressRabbit+=3;
               Message message=new Message();
               message.what=1;
               handler.sendMessage(message);
           }
        }).start();
    }

    private void runTurtle(){
        new Thread(()->{
            while(progressTurtle<=100 && progressRabbit<100){
                try{
                    Thread.sleep(100);
                }catch (InterruptedException e){
                    e.printStackTrace();;
                }
                progressTurtle+=1;
                Message message=new Message();
                message.what=2;
                handler.sendMessage(message);
            }
        }).start();
    }
}

File: Lab5_withKotlin/app/src/main/java/com/stella/lab5_withkotlin/MainActivity.kt
package com.stella.lab5_withkotlin

import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val btn = findViewById<Button>(R.id.button)
        btn.setOnClickListener {
            val dialog = AlertDialog.Builder(this@MainActivity)
            dialog.setTitle("請選擇功能")
            dialog.setMessage("請根據下方按鈕選擇顯示的物件")
            dialog.setNeutralButton(
                "取消"
            ) { dialog, i ->
                Toast.makeText(this@MainActivity, "dialog close", Toast.LENGTH_SHORT).show()
            }
            dialog.setNegativeButton(
                "自定義Toast"
            ) { dialog, i -> showToast() }
            dialog.setPositiveButton(
                "顯示list"
            ) { dialog, i -> showListDialog() }
            dialog.show()
        }
    }

    private fun showToast() {
        val toast = Toast(this@MainActivity)
        toast.setGravity(Gravity.CENTER, 0, 50)
        toast.duration = Toast.LENGTH_SHORT
        val inflater = layoutInflater
        val layout: View = inflater.inflate(R.layout.custom_toast
            , findViewById<View>(R.id.custom_toast_root) as ViewGroup?)
        toast.setView(layout)
        toast.show()
    }

    private fun showListDialog() {
        val list = arrayOf("message1", "message2", "message3", "message4", "message5")
        val dialog_list = AlertDialog.Builder(this@MainActivity)
        dialog_list.setTitle("使用list呈現")
        dialog_list.setItems(
            list
        ) { dialog, i ->
            Toast.makeText(this@MainActivity, "你選的是" + list[i], Toast.LENGTH_SHORT).show()
        }
        dialog_list.show()
    }
}